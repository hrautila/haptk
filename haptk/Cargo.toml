[package]
name = "haptk"
version = "0.3.0"
authors = ["Osma S. Rautila"]
edition = "2021"
description = "HAPTK - Haplotype analysis toolkit"
license = "GPL-3.0-only"
readme = "README.md"
homepage = "https://github.com/xosxos/haptk"
repository = "https://github.com/xosxos/haptk"
exclude = [
    "tests/*",
]


[lib]
crate-type = ["cdylib", "rlib"]
path = "src/lib.rs"
doctest = false

[[bin]]
name = "haptk"
path = "src/bin.rs"

[features]
default = ["clap"]
experimental = ["clap", "dep:fishers_exact"]
clap = ["dep:clap", "dep:anstyle"]

[dependencies]
alphanumeric-sort = "1.5"
bgzip = "0.3"
bio = "2.0"
niffler = "2.6"
ndarray = { version = "0.16", features = ["rayon", "blas", "serde"] }
ndarray-npy = { version = "0.9", default-features = false }
termion = "4.0"
color-eyre = "0.6"
rayon = "1.10"
serde = { version = "1.0", features = ["derive"] }
rand = "0.8.5"
svg = "0.13"
csv = "1.3"
itertools = "0.13"
serde_json = "1.0"
tracing = "0.1"
tracing-appender = "0.2"
indexmap = {version = "2.6", features = ["serde"]}
tracing-subscriber = { version = "0.3", default-features = false, features = ["std", "fmt", "time"] }
time = { version = "0.3", default-features = false, features = ["local-offset"] }
petgraph = { version = "0.6", default-features = false, features = ["serde-1"] }
rust-htslib = { version = "0.44", default-features = false, features = ['bzip2'] }
polars = { version = "=0.33.2", default-features = false, features = ["lazy", "is_in", "csv", "fmt", "serde"]}
image = { version = "0.25", default-features = false, features = ['png'] }
statrs = { version = "0.16", default-features = false }
chrono = { version = "0.4.38" }

# Cli
clap = { version = "=4.4.18", optional = true, features = ["derive"] }
anstyle = { version = "1.0.7", optional = true }

# Genome-wide
fishers_exact = { version="1.0.1", optional=true }


[dev-dependencies]
insta = { version = "1.26.0", default-features = false, features = ["yaml"] }



