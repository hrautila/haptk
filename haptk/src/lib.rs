// HATK - Haplotype analysis toolkit
// Copyright (C) 2022  Osma S. Rautila
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

pub mod libs;
pub use libs::{args, error, io, read_vcf, stats, structs, utils};

#[cfg(feature = "clap")]
pub use libs::clap;

/// Graphs implemented in Rust, HST graph will be replaced by the Python implementation
pub mod graphs;

/// HAPTK commands
pub mod subcommands;
