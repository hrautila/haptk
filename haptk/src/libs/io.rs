use std::collections::BTreeMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::{Path, PathBuf};

use color_eyre::{
    eyre::{eyre, WrapErr},
    Result,
};
use polars::prelude::*;

use crate::core::{get_csv_reader, get_input, get_tsv_reader};
use crate::structs::HapVariant;

pub fn read_variable_data_file(path: PathBuf) -> Result<DataFrame> {
    let df = CsvReader::from_path(path)?
        .with_null_values(Some(NullValues::AllColumnsSingle("NA".to_string())))
        .infer_schema(Some(500))
        .has_header(true)
        .finish()?;

    Ok(df)
}

#[derive(Debug, Clone, serde::Deserialize)]
struct RecombinationRow<'a> {
    _chr: &'a str,
    pos: u64,
    _rate: f32,
    cm: f32,
}

pub fn read_recombination_file(path: PathBuf) -> Result<BTreeMap<u64, f32>> {
    let mut rdr = get_tsv_reader(get_input(Some(path))?, false);
    let mut rates = BTreeMap::new();

    let mut last_cm = 0.0;
    let mut last_pos = 0;
    for line in rdr.records() {
        let record = line?;
        let row: RecombinationRow = record.deserialize(None).
            wrap_err(eyre!("Make sure no headers are present and that the recombination file is in order chr,pos,rate,cm. If the issue is not fixed, you have an invalid field in the file."))?;
        if last_cm > row.cm {
            return Err(eyre!(
                "Recombination rates file is not sorted for centimorgans"
            ));
        }
        if last_pos > row.pos {
            return Err(eyre!(
                "Recombination rates file is not sorted for positions"
            ));
        }
        last_cm = row.cm;
        last_pos = row.pos;
        rates.insert(row.pos, row.cm);
    }

    Ok(rates)
}

pub fn read_haplotype_file(ht_path: PathBuf) -> Result<Vec<HapVariant>> {
    let input = get_input(Some(ht_path))?;
    let mut rdr = get_csv_reader(input);

    let mut variants: Vec<HapVariant> = vec![];

    for line in rdr.records() {
        let record = line?;
        let variant: HapVariant = record.deserialize(None)?;
        if let Some(latest) = variants.last() {
            if latest.pos > variant.pos {
                return Err(eyre!(
                    "The haplotype file is not sorted by position. {} is larger than {}",
                    latest,
                    variant
                ));
            }
        }
        variants.push(variant);
    }

    Ok(variants)
}

pub fn read_lines<P>(filename: P) -> Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    // let file = File::open(filename)?;
    let name = filename.as_ref().display();
    let file = match File::open(&filename) {
        Ok(x) => x,
        Err(err) => {
            let msg = format!("failed to open {name}: {err}");
            return Err(std::io::Error::new(std::io::ErrorKind::NotFound, msg))?;
        }
    };
    Ok(io::BufReader::new(file).lines())
}

pub fn read_sample_ids(path: &Option<PathBuf>) -> Result<Option<Vec<String>>> {
    match path {
        Some(path) => {
            let mut samples = vec![];

            for line in read_lines(path)?.flatten() {
                let line = line.trim();
                samples.push(line.to_string());
            }
            Ok(Some(samples))
        }
        None => Ok(None),
    }
}
