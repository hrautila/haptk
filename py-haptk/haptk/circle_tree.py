from ete3 import TreeStyle, TextFace, CircleFace,  NodeStyle, add_face_to_node

from haptk import utils

def circle_tree_style():
    ts = TreeStyle()

    ts.branch_vertical_margin = 10
    ts.show_leaf_name = False
    ts.show_scale = False

    ts.mode = "c"
    ts.arc_start = 0 # 0 degrees = 15:00 o'clock
    ts.arc_span = 355
    ts.min_leaf_separation = 0
    ts.root_opening_factor = 0
    ts.optimal_scale_level = "mid"
    ts.allow_face_overlap = True

    return ts

def is_maj_up_to(hst, node, value):
    iter_node = node
    # print(f"checking {iter_node.name} for majority")
    for _ in range(0, value):
        if iter_node.name == "0":
            return True

        if not iter_node.up.name == "0":
            # print("input node:", iter_node.up)
            is_big = is_majority_node(hst, iter_node.up)
            if is_big == False:
                # print(f"run {_}: {iter_node.up.name} is not big")
                return False

        iter_node = iter_node.up
            
    return True

def is_majority_node(hst, node):
    parent = node.up
    input_node_data = hst.get_node_data(int(node.name))
    value = len(input_node_data["indexes"])
    # print(f"input node {node.name} has {value} indexes")

    siblings = {}
    for child in parent.children:
        data = hst.get_node_data(int(child.name))
        if not child.name == node.name:
            siblings[child.name] = len(data["indexes"])


    # print(value, siblings.values())
    for v in siblings.values():
        if value < v: 
            return False
  

def get_children_idx_amounts(hst, node):
    return [len(hst.get_node_data(c.name)["indexes"]) for c in node.children]

def draw_circle_tree(hst, t, output, samples, samples_to_tag, colors):
    style = NodeStyle()
    style["size"] = 0

    narea_color = "#DAF7A6"
    larea_color = "#FFB6C1"

    spent_parents = []
    spent_grand_parents = []

    for n in t.traverse():
        n.set_style(style)
        
        if not n.is_leaf():
            indexes = hst.get_node_indexes(n.name)
            label = len(indexes)
            F = TextFace(label, tight_text=True, penwidth=30, fsize=8)
            F.rotation = 270
            n.add_face(F, column=0)

            count = 0
            for i in get_children_idx_amounts(hst, n):
                if i > 80:
                    count += 1

            if count > 1:
                haplotype = hst.get_node_haplotype(n.name)
                F = TextFace(f"{len(haplotype)}", fsize=30, tight_text=True, fgcolor="red")
                # F.rotation = 270
                n.add_face(F, column=0)
                
                

        if n.is_leaf():
            if samples_to_tag == []:
                n.set_style(utils.return_node_style("#000", 0))
                n.img_style["bgcolor"] = "#FFF"

            for (color_i, sample_list) in enumerate(samples_to_tag):
                results = [i for i in indexes if samples[i] in sample_list]
                if results:
                    n.set_style(utils.return_node_style(colors[color_i], 0))
                    n.img_style["bgcolor"] = colors[color_i]

            if n.up:
                parent = n.up
                parent_indexes = hst.get_node_indexes(parent.name)

                if len(parent_indexes) == 2:
                    if parent in spent_parents:
                        if is_maj_up_to(hst, n, 14):
                            # print(f"{n.name} is_maj")

                            haplotype = hst.get_node_haplotype(n.name)
                            start_idx = hst.get_node_start_idx(n.name)
                            stop_idx = hst.get_node_stop_idx(n.name)
                            start = hst.coords[start_idx]
                            stop = hst.coords[stop_idx]
                            # print(start, stop)
                            bp = stop['pos'] - start['pos']
                            F = TextFace(f"{len(haplotype)}, {bp}", fsize=30, fgcolor="white")
                            # F.margin_left = 50
                            n.add_face(F, column=0)

                            # n.set_style(utils.return_node_style("white", 0))
                            # n.img_style["bgcolor"] = "white"
                    else:
                        spent_parents.append(parent)


    return t

