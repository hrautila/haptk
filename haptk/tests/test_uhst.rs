mod common;

#[cfg(test)]
#[cfg(feature = "clap")]
mod test_uhst {
    use super::*;

    use color_eyre::Result;

    use std::io::Write;
    use std::path::PathBuf;

    use haptk::{
        args::{Selection, StandardArgs},
        read_vcf::read_vcf_to_matrix,
        structs::Coord,
        subcommands::uhst_shard::{self, LocDirection},
    };

    #[test]
    fn matrix_to_graph() -> Result<()> {
        let args = StandardArgs {
            file: PathBuf::from(common::TEST_VCF),
            ..Default::default()
        };
        let mut vcf = read_vcf_to_matrix(&args, "chr9", 32, None, None, false).unwrap();

        let coord = Coord {
            contig: String::from("chr9"),
            reference: String::from("G"),
            alt: String::from("T"),
            pos: 32,
        };

        let g = uhst_shard::construct_uhst(&mut vcf, &LocDirection::Left, &coord, 1, false)?;
        let mut f = std::fs::File::create("tests/results/test.dot").unwrap();
        f.write_all(format!("{}", petgraph::dot::Dot::new(&g)).as_bytes())
            .unwrap();
        Ok(())
    }

    //--- Binary tests

    #[test]
    fn uhst_all() {
        run_uhst(Selection::All);

        let res = std::fs::read_to_string("tests/results/uhst_mbah.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype.csv").unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_ref() {
        run_uhst(Selection::OnlyRefs);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_refs.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_refs.csv")
            .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_alt() {
        run_uhst(Selection::OnlyAlts);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_alts.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res = std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_alts.csv")
            .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    #[test]
    fn uhst_only_longest() {
        run_uhst(Selection::OnlyLongest);

        let res = std::fs::read_to_string("tests/results/uhst_mbah_only_longest.csv").unwrap();
        insta::assert_yaml_snapshot!(res);

        let res =
            std::fs::read_to_string("tests/results/uhst_shared_core_haplotype_only_longest.csv")
                .unwrap();
        insta::assert_yaml_snapshot!(res);
    }

    fn run_uhst(selection: Selection) {
        let args = common::clap_standard_args(selection);

        let cmd = haptk::clap::SubCommand::Uhst {
            args,
            log_and_verbosity: crate::common::silent_verbosity(),
            threads: 8,
            min_size: 1,
            publish: false,
            sharded: false,
        };
        haptk::clap::run_cmd(cmd).unwrap();
    }
}
