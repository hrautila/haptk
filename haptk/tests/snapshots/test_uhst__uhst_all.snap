---
source: haptk/tests/test_uhst.rs
expression: res
---
"<svg style=\"background-color:white\" viewBox=\"0 0 20000 14000\" xmlns=\"http://www.w3.org/2000/svg\">\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"10000\" y=\"280\">\n28\n</text>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"9657.143\" y=\"1165\">\n27\n</text>\n<path d=\"M10241.5,317.8 L10241.5,1123 L9919.643,1123\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"19257.143\" y=\"1165\">\n1\n</text>\n<path d=\"M10241.5,317.8 L10241.5,1123 L19236.143,1123\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"9314.285\" y=\"1607.5\">\n26\n</text>\n<path d=\"M9898.643,1202.8 L9898.643,1565.5 L9576.785,1565.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"18571.428\" y=\"1607.5\">\n1\n</text>\n<path d=\"M9898.643,1202.8 L9898.643,1565.5 L18550.428,1565.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"8971.428\" y=\"2050\">\n25\n</text>\n<path d=\"M9555.785,1645.3 L9555.785,2008 L9233.928,2008\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"17885.713\" y=\"2050\">\n1\n</text>\n<path d=\"M9555.785,1645.3 L9555.785,2008 L17864.713,2008\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"8628.57\" y=\"2492.5\">\n24\n</text>\n<path d=\"M9212.928,2087.8 L9212.928,2450.5 L8891.07,2450.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"17199.998\" y=\"2492.5\">\n1\n</text>\n<path d=\"M9212.928,2087.8 L9212.928,2450.5 L17178.998,2450.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"8285.713\" y=\"2935\">\n23\n</text>\n<path d=\"M8870.07,2530.3 L8870.07,2893 L8548.213,2893\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"16514.283\" y=\"2935\">\n1\n</text>\n<path d=\"M8870.07,2530.3 L8870.07,2893 L16493.283,2893\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7942.856\" y=\"3377.5\">\n22\n</text>\n<path d=\"M8527.213,2972.8 L8527.213,3335.5 L8205.355,3335.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"15828.569\" y=\"3377.5\">\n1\n</text>\n<path d=\"M8527.213,2972.8 L8527.213,3335.5 L15807.569,3335.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7599.9985\" y=\"3820\">\n21\n</text>\n<path d=\"M8184.356,3415.3 L8184.356,3778 L7862.4985,3778\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"15142.8545\" y=\"3820\">\n1\n</text>\n<path d=\"M8184.356,3415.3 L8184.356,3778 L15121.8545,3778\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"7257.1416\" y=\"4262.5\">\n20\n</text>\n<path d=\"M7841.4985,3857.8 L7841.4985,4220.5 L7519.6416,4220.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"14457.141\" y=\"4262.5\">\n1\n</text>\n<path d=\"M7841.4985,3857.8 L7841.4985,4220.5 L14436.141,4220.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"6914.284\" y=\"4705\">\n19\n</text>\n<path d=\"M7498.6416,4300.3 L7498.6416,4663 L7176.784,4663\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"13771.426\" y=\"4705\">\n1\n</text>\n<path d=\"M7498.6416,4300.3 L7498.6416,4663 L13750.426,4663\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"6571.4272\" y=\"5147.5\">\n18\n</text>\n<path d=\"M7155.784,4742.8 L7155.784,5105.5 L6833.9272,5105.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"13085.711\" y=\"5147.5\">\n1\n</text>\n<path d=\"M7155.784,4742.8 L7155.784,5105.5 L13064.711,5105.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"6228.5703\" y=\"5590\">\n17\n</text>\n<path d=\"M6812.9272,5185.3 L6812.9272,5548 L6491.0703,5548\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"12399.998\" y=\"5590\">\n1\n</text>\n<path d=\"M6812.9272,5185.3 L6812.9272,5548 L12378.998,5548\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"5885.7134\" y=\"6032.5\">\n16\n</text>\n<path d=\"M6470.0703,5627.8 L6470.0703,5990.5 L6148.2134,5990.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"11714.283\" y=\"6032.5\">\n1\n</text>\n<path d=\"M6470.0703,5627.8 L6470.0703,5990.5 L11693.283,5990.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"5542.8564\" y=\"6475\">\n15\n</text>\n<path d=\"M6127.2134,6070.3 L6127.2134,6433 L5805.3564,6433\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"11028.57\" y=\"6475\">\n1\n</text>\n<path d=\"M6127.2134,6070.3 L6127.2134,6433 L11007.57,6433\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"5199.9995\" y=\"6917.5\">\n14\n</text>\n<path d=\"M5784.3564,6512.8 L5784.3564,6875.5 L5462.4995,6875.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"10342.855\" y=\"6917.5\">\n1\n</text>\n<path d=\"M5784.3564,6512.8 L5784.3564,6875.5 L10321.855,6875.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4857.142\" y=\"7360\">\n13\n</text>\n<path d=\"M5441.4995,6955.3 L5441.4995,7318 L5119.642,7318\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"9657.142\" y=\"7360\">\n1\n</text>\n<path d=\"M5441.4995,6955.3 L5441.4995,7318 L9636.142,7318\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4514.2847\" y=\"7802.5\">\n12\n</text>\n<path d=\"M5098.642,7397.8 L5098.642,7760.5 L4776.7847,7760.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"8971.427\" y=\"7802.5\">\n1\n</text>\n<path d=\"M5098.642,7397.8 L5098.642,7760.5 L8950.427,7760.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"4171.4277\" y=\"8245\">\n11\n</text>\n<path d=\"M4755.7847,7840.3 L4755.7847,8203 L4433.9277,8203\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"8285.713\" y=\"8245\">\n1\n</text>\n<path d=\"M4755.7847,7840.3 L4755.7847,8203 L8264.713,8203\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"3828.5706\" y=\"8687.5\">\n10\n</text>\n<path d=\"M4412.9277,8282.8 L4412.9277,8645.5 L4091.0706,8645.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"7599.998\" y=\"8687.5\">\n1\n</text>\n<path d=\"M4412.9277,8282.8 L4412.9277,8645.5 L7578.998,8645.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"3485.7134\" y=\"9130\">\n9\n</text>\n<path d=\"M4070.0706,8725.3 L4070.0706,9088 L3632.7134,9088\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"6914.284\" y=\"9130\">\n1\n</text>\n<path d=\"M4070.0706,8725.3 L4070.0706,9088 L6893.284,9088\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"3142.8564\" y=\"9572.5\">\n8\n</text>\n<path d=\"M3590.7134,9167.8 L3590.7134,9530.5 L3289.8564,9530.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"6228.57\" y=\"9572.5\">\n1\n</text>\n<path d=\"M3590.7134,9167.8 L3590.7134,9530.5 L6207.57,9530.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2799.9995\" y=\"10015\">\n7\n</text>\n<path d=\"M3247.8564,9610.3 L3247.8564,9973 L2946.9995,9973\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"5542.856\" y=\"10015\">\n1\n</text>\n<path d=\"M3247.8564,9610.3 L3247.8564,9973 L5521.856,9973\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2457.1423\" y=\"10457.5\">\n6\n</text>\n<path d=\"M2904.9995,10052.8 L2904.9995,10415.5 L2604.1423,10415.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"4857.1416\" y=\"10457.5\">\n1\n</text>\n<path d=\"M2904.9995,10052.8 L2904.9995,10415.5 L4836.1416,10415.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"2114.2852\" y=\"10900\">\n5\n</text>\n<path d=\"M2562.1423,10495.3 L2562.1423,10858 L2261.2852,10858\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"4171.4277\" y=\"10900\">\n1\n</text>\n<path d=\"M2562.1423,10495.3 L2562.1423,10858 L4150.4277,10858\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"1771.4282\" y=\"11342.5\">\n4\n</text>\n<path d=\"M2219.2852,10937.8 L2219.2852,11300.5 L1918.4282,11300.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"3485.7134\" y=\"11342.5\">\n1\n</text>\n<path d=\"M2219.2852,10937.8 L2219.2852,11300.5 L3464.7134,11300.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"1428.5712\" y=\"11785\">\n3\n</text>\n<path d=\"M1876.4282,11380.3 L1876.4282,11743 L1575.5712,11743\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"2799.9995\" y=\"11785\">\n1\n</text>\n<path d=\"M1876.4282,11380.3 L1876.4282,11743 L2778.9995,11743\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"1085.7141\" y=\"12227.5\">\n2\n</text>\n<path d=\"M1533.5712,11822.8 L1533.5712,12185.5 L1232.7141,12185.5\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"2114.2852\" y=\"12227.5\">\n1\n</text>\n<path d=\"M1533.5712,11822.8 L1533.5712,12185.5 L2093.2852,12185.5\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"1\" x=\"742.85706\" y=\"12670\">\n1\n</text>\n<path d=\"M1190.7141,12265.3 L1190.7141,12628 L889.85706,12628\" fill=\"none\" opacity=\"1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"black\" font-size=\"210px\" font-weight=\"bold\" opacity=\"0.1\" x=\"1428.5712\" y=\"12670\">\n1\n</text>\n<path d=\"M1190.7141,12265.3 L1190.7141,12628 L1407.5712,12628\" fill=\"none\" opacity=\"0.1\" stroke=\"black\" stroke-width=\"20\"/>\n<text fill=\"#ff2f8e\" font-size=\"147px\" font-weight=\"bold\" opacity=\"1\" x=\"742.85706\" y=\"12880\">\n0.00 mb\n</text>\n<text fill=\"#66df48\" font-size=\"147px\" font-weight=\"bold\" opacity=\"1\" x=\"742.85706\" y=\"13090\">\nn=28\n</text>\n<text fill=\"#6a77dd\" font-size=\"147px\" font-weight=\"bold\" opacity=\"1\" x=\"742.85706\" y=\"13300\">\n62.69 y\n</text>\n</svg>"

